py/modsys.c: Add sys._exc_traceback.

DavidEGrayson opened the pull request:
This new function makes it easier to provide custom formatting of exception stack traces, for example to make them fit on a tiny 16x8-character display.  (Without this, I think you'd have to call `sys.print_exception`, somehow get the output as a string, and then scrape information from that string.)

I'm thinking of this as an experimental function that exposes internal details of MicroPython and therefore might change in the future.  That's why it has the underscore in the name.

Here is an example of its output:

    ['test.py', 4, 'foo', 'test.py', 7, '<module>']

The regular output from `sys.print_exception` for the same exception is:

```text
Traceback (most recent call last):
  File "test.py", line 7, in <module>
  File "test.py", line 4, in foo
ZeroDivisionError: divide by zero
```

Both outputs were generated by the following script:

```text
import sys

def foo():
  1/0

try:
  foo()
except Exception as e:
  sys.print_exception(e)
  print(sys._exc_traceback(e))
 ```

Is this the right approach for getting info about exception stack traces?   I'd be happy to change the name, change the output format, add tests, or add documentation.

dpgeorge commented:
First thing to ask is how would one do this in CPython?

Then we need to consider if it's worth the cost (code size) and whether it should be guarded by a config macro.  I would say yes it should be, but then at what level would it be enabled, eg `MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES`.

DavidEGrayson commented:
You can get this info in CPython is like this:

```py
import traceback

def foo():
    1/0

try:
    foo()
except Exception as e:
    tb = e.__traceback__
    while tb is not None:
        print("====")
        print("  filename: " + str(tb.tb_frame.f_code.co_filename))
        print("  line: " + str(tb.tb_lineno))
        print("  name: " + str(tb.tb_frame.f_code.co_name))
        tb = tb.tb_next
```

This involves at least three different classes: `traceback`, `frame`, and `code`.  The latter two contain quite a few things:

```text
dir(tb): ['tb_frame', 'tb_lasti', 'tb_lineno', 'tb_next']
dir(frame): ['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'clear', 'f_back', 'f_builtins', 'f_code', 'f_globals', 'f_lasti', 'f_lineno', 'f_locals', 'f_trace', 'f_trace_lines', 'f_trace_opcodes']
dir(code): ['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'co_argcount', 'co_cellvars', 'co_code', 'co_consts', 'co_filename', 'co_firstlineno', 'co_flags', 'co_freevars', 'co_kwonlyargcount', 'co_lines', 'co_linetable', 'co_lnotab', 'co_name', 'co_names', 'co_nlocals', 'co_posonlyargcount', 'co_stacksize', 'co_varnames', 'replace']
```

The Python standard library also comes with a [traceback module](https://github.com/python/cpython/blob/3.11/Lib/traceback.py) that introduces more classes.  MicroPython implements a [minimal traceback module](https://github.com/micropython/micropython-lib/blob/master/python-stdlib/traceback/traceback.py).

DavidEGrayson commented:
I changed it so that this function is only enabled by default if `MICROPY_CONFIG_ROM_LEVEL_EXTRA_FEATURES` is 1.  (I'm glad to see this provides it for the RP2 boards.)  I also added documentation.

dpgeorge commented:
Thanks for keeping this PR alive.  I can see how it would be useful.

To reduce code size I suggest implementing the functionality as an attribute of the exception type, rather than as a new function.  So it would go in `mp_obj_exception_attr()`.  To keep code size further down, it could reuse the existing `__traceback__` attribute name.  So the example from above would be:
```py
import sys

def foo():
  1/0

try:
  foo()
except Exception as e:
  sys.print_exception(e)
  print(e.__traceback__)
```

Now, that's not fully compatible with CPython, but it's a little bit closer than having a separate `sys._exc_traceback()` function.  We can say that if you want compatibility with CPython then you must consider `e.__traceback__` as an opaque (or implementation defined) value, and then use the `traceback` module to convert that opaque value to something usable.  Eg `traceback.extract_tb(e.__traceback__)` would work the same in CPython and MicroPython, even though the underlying value of `e.__traceback__` was different.

And then if you don't care about CPython compatibility and you want to efficiently get the traceback data, you can use the value of `e.__traceback__` directly (although maybe we should document it as opaque and subject to change).

DavidEGrayson committed:
py/objexcept: Add __traceback__ attribute getter.

This makes it easier to provide custom formatting of exception stack
traces, e.g. to make them fit on a tiny display.

Signed-off-by: David (Pololu) <dev-david@pololu.com>

dpgeorge commented:
@DavidEGrayson I'll leave this up to you, let me know after you've had a chance to get back to it.

dpgeorge commented on py/objexcept.c:
```diff
@@ -280,7 +280,23 @@ void mp_obj_exception_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+            size_t *src = &data[i * TRACEBACK_ENTRY_LEN];
+            mp_obj_t entry[3];
+            entry[0] = MP_OBJ_NEW_QSTR(src[0]);       // filename
+            entry[1] = MP_OBJ_NEW_SMALL_INT(src[1]);  // line number
```
If MICROPY_ENABLE_SOURCE_LINE is disabled then this number is always 1.  Maybe worth storing 0 (or -1?) instead, to indicate that line numbers are unavailable?

dpgeorge commented on py/objexcept.c:
```diff
@@ -280,7 +280,23 @@ void mp_obj_exception_attr(mp_obj_t self_in, qstr attr, mp_obj_t *dest) {
+            mp_obj_t entry[3];
+            entry[0] = MP_OBJ_NEW_QSTR(src[0]);       // filename
+            entry[1] = MP_OBJ_NEW_SMALL_INT(src[1]);  // line number
+            entry[2] = MP_OBJ_NEW_QSTR(src[2]);       // block
```
`src[2]` may be `MP_QSTRnull` if the block name is unknown, which shouldn't then be converted to a qstr object.

Not sure what the best way to handle this is.  Maybe change it so `MP_QSTR_` is used instead of `MP_QSTRnull` when the block name is unknown.

dpgeorge reviewed (commented):
Regarding replacing `sys.print_exception(exc)` with this new `exc.__traceback__` feature: that might not work very well because `sys.print_exception` has a lot of nice features that aren't easily available with `exc.__traceback__`, such as errno name printing and nice formatting of the exception type and arguments.

I guess you could write Python to emulate that pretty printing.  Alternatively, could add another attribute to exceptions that return a pretty exception string, eg `exc.__str__` (to reuse a qstr).
